### PR TITLE
FeatureParsers, CLI, Comparators and Main class

### PR DESCRIPTION
**Pull request by MarchenkoVladislav**

### CODE REVIEW

File: bio-common/src/main/kotlin/bio/relatives/common/cli/command/ComparisonCommand.kt [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+                RunnerCtx(comparatorType,
+                        Paths.get(pathToFeatureFile!!),
+                        Paths.get(pathToFatherFile!!),
+                        Paths.get(pathToMotherFile!!),
+                        Paths.get(pathToSonFile!!)
+                )
+        )
+    }
+
+    companion object {
```
Comment (SHvatov) [inline bio-common/src/main/kotlin/bio/relatives/common/cli/command/ComparisonCommand.kt:67]: private -> companion object

File: bio-common/src/main/kotlin/bio/relatives/common/utils/MathUtils.kt [hunk 2]
```diff
@@ -40,5 +41,14 @@ fun <T : Number> calculateAverageQuality(values: List<T>): Double {
...
 fun <T : Number> calculateAdditionRelativeErrorRate(valueToError: Map<T, T>): Double {
     return valueToError.map { it.value.toDouble() }.sum() /
-        valueToError.map { it.key.toDouble() }.sum()
+            valueToError.map { it.key.toDouble() }.sum()
 }
+
+/**
+ * Rounds a [valueToRound] to [placesNum] decimal places
+ */
+fun round(valueToRound: Double, placesNum: Byte): Double =
```
[the remark this answers is not in the document]
Comment (SHvatov) [reply bio-common/src/main/kotlin/bio/relatives/common/utils/MathUtils.kt:50]: if you want to prettify the output, then do it after the comparison in the output of the analyzer

File: bio-common/src/main/kotlin/bio/relatives/common/utils/NucleotideSequenceUtils.kt [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    val secondGenome = StringBuilder()
+    for (i in 0 until min(first.length, second.length)) {
+
+        if (!isUnknownNucleotide(first[i]) && !isUnknownNucleotide(second[i])) {
+            firstGenome.append(first[i])
+            secondGenome.append(second[i])
+        }
+    }
+
+    if (firstGenome.length < secondGenome.length) {
```
Comment (SHvatov) [inline bio-common/src/main/kotlin/bio/relatives/common/utils/NucleotideSequenceUtils.kt:25]: no need for that at the moment

File: bio-distance/src/main/kotlin/bio/relatives/bio/distance/comparator/BioDistanceGenomeComparatorAlgorithm.kt [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+import bio.relatives.common.model.Region
+import bio.relatives.common.utils.getNormalizedAlignments
+import bio.relatives.common.utils.round
+import java.lang.Integer.max
+import java.lang.Integer.min
+
+/**
+ * @author Created by Vladislav Marchenko on 30.01.2021
+ */
+class BioDistanceGenomeComparatorAlgorithm : GenomeComparatorAlgorithm {
```
Comment (SHvatov) [inline bio-distance/src/main/kotlin/bio/relatives/bio/distance/comparator/BioDistanceGenomeComparatorAlgorithm.kt:15]: LevenshteinDistanceAlgorithm

File: bio-distance/src/main/kotlin/bio/relatives/bio/distance/comparator/BioDistanceGenomeComparatorAlgorithm.kt [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+    override fun compare(left: Region, right: Region, feature: Feature): ComparisonResult.ComparisonAlgorithmResult {
+
+        val temp: Pair<String, String> = getNormalizedAlignments(left.sequence, right.sequence)
+
+        val f: String = temp.first
```
Comment (SHvatov) [inline bio-distance/src/main/kotlin/bio/relatives/bio/distance/comparator/BioDistanceGenomeComparatorAlgorithm.kt:21]: val (f, s) = temp

File: bio-distance/src/main/kotlin/bio/relatives/bio/distance/comparator/BioDistanceGenomeComparatorAlgorithm.kt [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+            for (k in 1 until s.length + 1) {
+                current[k] = min(min(current[k - 1] + 1, table[k] + 1), min(table[k] + 1, table[k - 1] + if (f[l - 1] == s[k - 1]) 0 else 1))
+            }
+            table = current.copyOf(current.size)
+        }
+
+        val difference = current[s.length]
+        val length = max(f.length, s.length)
+        val similarityPercent = round(100.0 - difference.toDouble() / length.toDouble() * 100.0, 2)
+        val errorRate = round(100 - arrayOf(*left.qualities, *right.qualities).average(), 2)
```
Comment (SHvatov) [inline bio-distance/src/main/kotlin/bio/relatives/bio/distance/comparator/BioDistanceGenomeComparatorAlgorithm.kt:43]: use our functions

File: bio-distance/src/main/kotlin/bio/relatives/bio/distance/model/BEDFeature.kt [hunk 7]
```diff
@@ -0,0 +1,13 @@
+package bio.relatives.bio.distance.model
+
+import bio.relatives.common.model.Feature
+
+/**
+ * @author Created by Vladislav Marchenko on 01.02.2021
+ */
+class BEDFeature(
```
Comment (SHvatov) [inline bio-distance/src/main/kotlin/bio/relatives/bio/distance/model/BEDFeature.kt:8]: BedFeature

File: bio-distance/src/main/kotlin/bio/relatives/bio/distance/parser/BEDFeatureParser.kt [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+import bio.relatives.common.parser.FeatureParser.Companion.MAX_FEATURE_SIZE
+import bio.relatives.common.utils.isValid
+import org.springframework.stereotype.Component
+import java.nio.file.Path
+
+/**
+ * @author Created by Vladislav Marchenko on 02.02.2021
+ */
+@Component
+class BEDFeatureParser : FeatureParser {
```
Comment (SHvatov) [inline bio-distance/src/main/kotlin/bio/relatives/bio/distance/parser/BEDFeatureParser.kt:16]: see above about naming + abstract class with common impl

File: bio-distance/src/main/kotlin/bio/relatives/bio/distance/parser/BEDFeatureParser.kt [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+        featureFilePath.isValid(requiredExtension = FEATURE_FILE_EXTENSION)
+
+        val res = mutableListOf<BEDFeature>()
+
+        featureFilePath.toFile().forEachLine {
+            run {
+                if (it != "/n" && !it.startsWith("#")) {
+                    val rows: Array<String> = it.split("[\\s+\\t+]".toRegex()).filter { it.isNotBlank() }.toTypedArray()
+
+                    require(rows.size == 4) {
```
Comment (SHvatov) [inline bio-distance/src/main/kotlin/bio/relatives/bio/distance/parser/BEDFeatureParser.kt:27]: in `abstract fun process(rows: List<String>): Feature`

File: bio-distance/src/main/kotlin/bio/relatives/bio/distance/parser/BEDFeatureParser.kt [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+                        "Error occurred during reading from the file [$featureFilePath]: " +
+                                "incorrect number of rows in the table. Expected 4 (chrom, start, end, gene name), got ${rows.size}"
+                    }
+
+                    val chrom = rows[0]
+                    val gene = rows[3]
+                    val start = rows[1].toInt()
+                    val end = rows[2].toInt()
+
+                    val exonLen = end - start
```
Comment (SHvatov) [inline bio-distance/src/main/kotlin/bio/relatives/bio/distance/parser/BEDFeatureParser.kt:37]: in abstract

File: bio-str/src/main/kotlin/bio/relatives/bio/y/str/comparator/BioYStrGenomeComparatorAlgorithm.kt [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+import bio.relatives.common.utils.getNormalizedAlignments
+import bio.relatives.common.utils.round
+import java.lang.Integer.min
+import java.lang.Math.max
+import java.util.regex.Matcher
+
+/**
+ * @author Created by Vladislav Marchenko on 30.01.2021
+ */
+class BioYStrGenomeComparatorAlgorithm : GenomeComparatorAlgorithm {
```
Comment (SHvatov) [inline bio-str/src/main/kotlin/bio/relatives/bio/y/str/comparator/BioYStrGenomeComparatorAlgorithm.kt:17]: naming

File: bio-str/src/main/kotlin/bio/relatives/bio/y/str/model/MarkerFeature.kt [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+
+/**
+ * @author Created by Vladislav Marchenko on 02.02.2021
+ */
+class MarkerFeature(
+        override val gene: String,
+        override val chromosome: String,
+        override val start: Int,
+        override val end: Int,
+        override val repeatMotif: Pattern) : Feature, RepeatMotifAware
```
Comment (SHvatov) [inline bio-str/src/main/kotlin/bio/relatives/bio/y/str/model/MarkerFeature.kt:15]: \n

File: bio-str/src/main/kotlin/bio/relatives/bio/y/str/parser/MarkerFeatureParser.kt [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+ */
+@Component
+class MarkerFeatureParser : FeatureParser {
+    override fun parseFeatures(featureFilePath: Path): List<Feature> {
+        featureFilePath.isValid(requiredExtension = FEATURE_FILE_EXTENSION)
+
+        val res = mutableListOf<MarkerFeature>()
+
+        featureFilePath.toFile().forEachLine {
+            run {
```
Comment (SHvatov) [inline bio-str/src/main/kotlin/bio/relatives/bio/y/str/parser/MarkerFeatureParser.kt:24]: what for

File: bio-common/src/main/kotlin/bio/relatives/common/cli/runner/impl/RunnerImpl.kt [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+ */
+@Component
+class RunnerImpl @Autowired constructor(
+        private val genomeAssemblerFactory: GenomeAssemblerFactory,
+        private val genomeComparatorFactory: GenomeComparatorFactory,
+        private val comparisonResultsAnalyserFactory: ComparisonResultsAnalyserFactory
+) : Runner {
+
+    private val scope = CoroutineScope(
+            SupervisorJob() + CoroutineName("Iterator via comparator")
```
Comment (SHvatov) [inline bio-common/src/main/kotlin/bio/relatives/common/cli/runner/impl/RunnerImpl.kt]: name -> Runner

Comment (SHvatov) [reply bio-common/src/main/kotlin/bio/relatives/common/cli/runner/impl/RunnerImpl.kt]: see latest commit

**Review by SHvatov — CHANGES_REQUESTED** [review]
See comments

File: bio-common/src/main/kotlin/bio/relatives/common/cli/command/ComparisonCommand.kt [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+
+    private val pathToSonFile by option(
+            ArgType.String,
+            shortName = PATH_TO_SON_FILE_OPTION_SHORT_NAME,
+            description = PATH_TO_SON_FILE_OPTION_DESCRIPTION
+    )
+
+    override fun execute() {
+        runner.run(
+                RunnerCtx(comparatorType,
```
[the remark this answers is not in the document]
Comment (MarchenkoVladislav) [reply bio-common/src/main/kotlin/bio/relatives/common/cli/command/ComparisonCommand.kt]: complete

File: bio-common/src/main/kotlin/bio/relatives/common/cli/parser/ArgParserFactory.kt [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+
+import bio.relatives.common.cli.runner.Runner
+import kotlinx.cli.ArgParser
+
+/**
+ * @author Created by Vladislav Marchenko on 04.02.2021
+ */
+interface ArgParserFactory {
+    /**
+     * Creates [ArgParser]
```
[the remark this answers is not in the document]
Comment (MarchenkoVladislav) [reply bio-common/src/main/kotlin/bio/relatives/common/cli/parser/ArgParserFactory.kt]: complete

File: bio-common/src/main/kotlin/bio/relatives/common/comparator/GenomeComparatorAlgorithm.kt [hunk 17]
```diff
@@ -11,5 +12,5 @@ interface GenomeComparatorAlgorithm {
     /**
      * Compares provided [Region] instances and returns the result as [ComparisonAlgorithmResult].
      */
-    fun compare(left: Region, right: Region): ComparisonAlgorithmResult
+    fun compare(left: Region, right: Region, feature: Feature): ComparisonAlgorithmResult
```
[the remark this answers is not in the document]
Comment (MarchenkoVladislav) [reply bio-common/src/main/kotlin/bio/relatives/common/comparator/GenomeComparatorAlgorithm.kt]: complete

File: bio-common/src/main/kotlin/bio/relatives/common/model/RegionBatch.kt [hunk 18]
```diff
@@ -0,0 +1,10 @@
+package bio.relatives.common.model
+
+
+/**
+ * @author Created by Vladislav Marchenko on 03.02.2021
+ */
+data class RegionBatch(
+        val feature: Feature,
+        val payload: HashMap<RoleAware.Role, Region>
```
[the remark this answers is not in the document]
Comment (MarchenkoVladislav) [reply bio-common/src/main/kotlin/bio/relatives/common/model/RegionBatch.kt]: complete

File: bio-common/src/main/kotlin/bio/relatives/common/parser/FeatureParser.kt [hunk 19]
```diff
@@ -7,5 +7,20 @@ import java.nio.file.Path
...
+    /**
+     * Parses feature file along the path [featureFilePath] into list of [Feature]
+     */
     fun parseFeatures(featureFilePath: Path): List<Feature>
+
+    companion object {
+        /**
+         * Maximal feature size
+         */
+        protected const val MAX_FEATURE_SIZE = 120
```
[the remark this answers is not in the document]
Comment (MarchenkoVladislav) [reply bio-common/src/main/kotlin/bio/relatives/common/parser/FeatureParser.kt]: complete

File: bio-common/src/main/kotlin/bio/relatives/common/processor/AssemblyProcessor.kt [hunk 20]
```diff
@@ -39,16 +39,16 @@ class AssemblyProcessor(
...
                 parentScope.async {
-                    assemble(it, payload)
+                    assemble(it, batch)
                 }
             }
 
-            return assemblyResults
-                .map { it.await() }
-                .associateByTo(RegionBatch()) { it.role }
+            return RegionBatch(batch, assemblyResults
```
[the remark this answers is not in the document]
Comment (MarchenkoVladislav) [reply bio-common/src/main/kotlin/bio/relatives/common/processor/AssemblyProcessor.kt:49]: complete

File: bio-common/src/main/kotlin/bio/relatives/common/utils/MathUtils.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (MarchenkoVladislav) [reply bio-common/src/main/kotlin/bio/relatives/common/utils/MathUtils.kt:50]: Now it  is using in the analyzer for prettifying the output

File: bio-distance/src/main/kotlin/bio/relatives/bio/distance/comparator/BioDistanceGenomeComparatorAlgorithm.kt [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+            current[0] = l
+            for (k in 1 until s.length + 1) {
+                current[k] = min(min(current[k - 1] + 1, table[k] + 1), min(table[k] + 1, table[k - 1] + if (f[l - 1] == s[k - 1]) 0 else 1))
+            }
+            table = current.copyOf(current.size)
+        }
+
+        val difference = current[s.length]
+        val length = max(f.length, s.length)
+        val similarityPercent = round(100.0 - difference.toDouble() / length.toDouble() * 100.0, 2)
```
[the remark this answers is not in the document]
Comment (MarchenkoVladislav) [reply bio-distance/src/main/kotlin/bio/relatives/bio/distance/comparator/BioDistanceGenomeComparatorAlgorithm.kt]: complete

File: bio-common/src/main/kotlin/bio/relatives/common/cli/command/ComparisonCommand.kt (hunk 1 above)
Comment (MarchenkoVladislav) [reply bio-common/src/main/kotlin/bio/relatives/common/cli/command/ComparisonCommand.kt]: completed

File: bio-common/src/main/kotlin/bio/relatives/common/utils/NucleotideSequenceUtils.kt (hunk 3 above)
Comment (MarchenkoVladislav) [reply bio-common/src/main/kotlin/bio/relatives/common/utils/NucleotideSequenceUtils.kt]: completed

File: bio-distance/src/main/kotlin/bio/relatives/bio/distance/comparator/BioDistanceGenomeComparatorAlgorithm.kt (hunk 4 above)
Comment (MarchenkoVladislav) [reply bio-distance/src/main/kotlin/bio/relatives/bio/distance/comparator/BioDistanceGenomeComparatorAlgorithm.kt]: completed

File: bio-distance/src/main/kotlin/bio/relatives/bio/distance/comparator/BioDistanceGenomeComparatorAlgorithm.kt (hunk 5 above)
Comment (MarchenkoVladislav) [reply bio-distance/src/main/kotlin/bio/relatives/bio/distance/comparator/BioDistanceGenomeComparatorAlgorithm.kt]: completed

File: bio-distance/src/main/kotlin/bio/relatives/bio/distance/comparator/BioDistanceGenomeComparatorAlgorithm.kt (hunk 6 above)
Comment (MarchenkoVladislav) [reply bio-distance/src/main/kotlin/bio/relatives/bio/distance/comparator/BioDistanceGenomeComparatorAlgorithm.kt]: completed

File: bio-distance/src/main/kotlin/bio/relatives/bio/distance/model/BEDFeature.kt (hunk 7 above)
Comment (MarchenkoVladislav) [reply bio-distance/src/main/kotlin/bio/relatives/bio/distance/model/BEDFeature.kt]: completed

File: bio-distance/src/main/kotlin/bio/relatives/bio/distance/parser/BEDFeatureParser.kt (hunk 8 above)
Comment (MarchenkoVladislav) [reply bio-distance/src/main/kotlin/bio/relatives/bio/distance/parser/BEDFeatureParser.kt]: completed

File: bio-distance/src/main/kotlin/bio/relatives/bio/distance/parser/BEDFeatureParser.kt (hunk 9 above)
Comment (MarchenkoVladislav) [reply bio-distance/src/main/kotlin/bio/relatives/bio/distance/parser/BEDFeatureParser.kt]: completed

File: bio-distance/src/main/kotlin/bio/relatives/bio/distance/parser/BEDFeatureParser.kt (hunk 10 above)
Comment (MarchenkoVladislav) [reply bio-distance/src/main/kotlin/bio/relatives/bio/distance/parser/BEDFeatureParser.kt]: completed

File: bio-str/src/main/kotlin/bio/relatives/bio/y/str/comparator/BioYStrGenomeComparatorAlgorithm.kt (hunk 11 above)
Comment (MarchenkoVladislav) [reply bio-str/src/main/kotlin/bio/relatives/bio/y/str/comparator/BioYStrGenomeComparatorAlgorithm.kt]: completed

File: bio-str/src/main/kotlin/bio/relatives/bio/y/str/model/MarkerFeature.kt (hunk 12 above)
Comment (MarchenkoVladislav) [reply bio-str/src/main/kotlin/bio/relatives/bio/y/str/model/MarkerFeature.kt]: completed

File: bio-str/src/main/kotlin/bio/relatives/bio/y/str/parser/MarkerFeatureParser.kt (hunk 13 above)
Comment (MarchenkoVladislav) [reply bio-str/src/main/kotlin/bio/relatives/bio/y/str/parser/MarkerFeatureParser.kt]: completed

File: bio-common/src/main/kotlin/bio/relatives/common/processor/CompareProcessor.kt [hunk 22]
```diff
@@ -22,16 +18,16 @@ class CompareProcessor(
...
+                batch: RegionBatch
         ): ComparisonResult {
             val results =
                 mutableMapOf<ComparisonParticipants, Deferred<ComparisonAlgorithmResult>>()
             for (between in ComparisonParticipants.COMPARISON_PAIRS) {
-                val (first, second) = payload[between.firstRole] to payload[between.secondRole]
+                val (first, second) = batch.regionsByRole[between.firstRole] to batch.regionsByRole[between.secondRole]
                 if (first != null && second != null) {
                     results[between] = parentScope.async {
-                        compareCtx.algorithm.compare(first, second)
```
Comment (SHvatov) [inline bio-common/src/main/kotlin/bio/relatives/common/processor/CompareProcessor.kt:34]: Let's sync our formatting - at the moment I am using google coding guide lines, you can find xml file with config for idea on the internet

File: bio-common/src/main/kotlin/bio/relatives/common/utils/NucleotideSequenceUtils.kt [hunk 23]
```diff
@@ -0,0 +1,32 @@
...
+    val secondGenome = StringBuilder()
+    for (i in 0 until min(first.length, second.length)) {
+
+        if (!isUnknownNucleotide(first[i]) && !isUnknownNucleotide(second[i])) {
+            firstGenome.append(first[i])
+            secondGenome.append(second[i])
+        }
+    }
+
+    return Pair(firstGenome.toString(), secondGenome.toString())
```
Comment (SHvatov) [inline bio-common/src/main/kotlin/bio/relatives/common/utils/NucleotideSequenceUtils.kt:25]: Here you can use kotlin 'to' infix function

File: bio-distance/src/main/kotlin/bio/relatives/bio/distance/comparator/LevensteinDistanceAlgorithm.kt [hunk 24]
```diff
@@ -0,0 +1,47 @@
...
+/**
+ * @author Created by Vladislav Marchenko on 30.01.2021
+ */
+class LevensteinDistanceAlgorithm : GenomeComparatorAlgorithm {
+
+    override fun compare(feature: Feature, left: Region, right: Region): ComparisonResult.ComparisonAlgorithmResult {
+
+        val temp = getNormalizedAlignments(left.sequence, right.sequence)
+
+        val f = temp.first
```
Comment (SHvatov) [inline bio-distance/src/main/kotlin/bio/relatives/bio/distance/comparator/LevensteinDistanceAlgorithm.kt:21]: Deconstruction initialisation

File: bio-str/src/main/kotlin/bio/relatives/bio/y/str/model/MarkerFeature.kt (hunk 12 above)
Comment (SHvatov) [inline bio-str/src/main/kotlin/bio/relatives/bio/y/str/model/MarkerFeature.kt:15]: New line here

File: bio-str/src/main/kotlin/bio/relatives/bio/y/str/parser/MarkerFeatureParser.kt [hunk 25]
```diff
@@ -0,0 +1,53 @@
...
+        } else {
+            var tempStart = start
+            for (tempEnd in start..end) {
+                if (tempEnd - tempStart > MAX_FEATURE_SIZE) {
+                    res.add(MarkerFeature(chrom, gene, tempStart, tempEnd, repeatMotif))
+                    tempStart = tempEnd + 1
+                }
+            }
+            if (end - tempStart > 1) {
+                res.add(MarkerFeature(chrom, gene, tempStart, end, repeatMotif))
```
Comment (SHvatov) [inline bio-str/src/main/kotlin/bio/relatives/bio/y/str/parser/MarkerFeatureParser.kt:47]: Yeah, code duplication, move to abstract impl

**Review by SHvatov — CHANGES_REQUESTED** [review]
Please, check out latest comment

**Pull request merged**